### DISCUSSION TITLE
Better float window support for v0.0.80

### DISCUSSION THREAD

**Discussion by chemzqm** [neoclide/coc.nvim#2472] [Show and tell]
We've done some work with float window/popup support recently, but there're still more work remains.

New dialog features works on neovim >= 0.4.0 or vim >= 8.2.750

- Border and title on both vim and neovim, like new action menu:

  <img width="411" alt="Screen Shot 2020-10-17 at 12 24 55 PM" src="https://user-images.githubusercontent.com/251450/96328272-c659f600-1073-11eb-95e9-1e147d87d981.png">

  Type number to select or use `j` `k` `G` to navigate and `<CR>` for confirm, use `<esc>` to cancel.

- New input prompt avoid the need to move your eyes.

  <img width="413" alt="Screen Shot 2020-10-17 at 12 31 57 PM" src="https://user-images.githubusercontent.com/251450/96328360-c0b0e000-1074-11eb-9104-82e2d39b2766.png">

   You can use keys including `<c-u>` `<c-e>` `<c-a>` , it's using normal buffer on neovim and terminal buffer on vim, so the experience could be a bit differrent. 

- Functions for scroll float window and popups, checkout `:h coc#float#has_scroll()`, new functions have improved for:

  - Works on both vim and neovim.
  - Works on both all modes, including normal, insert and visual.
  - Scroll all scrollable float windows/popups at the same time.
  - Works with wrapped buffer.

- Configuration `signature.floatMaxWidth` changed to `signature.maxWindowWidth` with detault 80.

- Create diaglog support:

   Use vim's api like: `call coc#float#create_dialog(['你好', 'vim'], #{close: 0, title: 'hello world', buttons: ['yes', 'no']})`
   Or use `window.showDiaglog(config)`

   <img width="168" alt="Screen Shot 2020-10-31 at 3 24 28 PM" src="https://user-images.githubusercontent.com/251450/97773750-698b2f00-1b8d-11eb-8d00-5fda250fba1e.png">
    
    Buttons accept mouse click on neovim only, vim users have to type specific keys.

- Notifications:

  <img width="349" alt="Screen Shot 2020-11-28 at 8 51 40 PM" src="https://user-images.githubusercontent.com/251450/100516232-76676680-31bd-11eb-9886-f1146cbe2136.png">

- Progress notification:

    <img width="329" alt="Screen Shot 2020-12-02 at 7 01 22 PM" src="https://user-images.githubusercontent.com/251450/100864378-dff2b800-34d0-11eb-9655-bbee92ba831d.png">

Plans:

- [x] Support border config as numbers on neovim (like vim).
- [x] Support scrollbar and close button on neovim (like vim).
- [x] Improve calculation of float window/popup config by use `strdisplaywidth`
- [x] Use float window for prompt confirm on neovim.
- [x] Support float window/popup with buttons, like a dialog.
- [x] Support dialog for multiple selection.
- [x] Notifications support.
- [x] Progress float window/popup support.

Some internal APis would be changed and you'll probably have something wired, give us feedback.

**Comment by oblitum** [comment]
> Works on both insert and normal mode.

Oh, I only noticed that now. I've recently merged a pull to have insert mode scroll in nvim, if this covers that then please just remove the old one. I'll try this new one now.

**Comment by oblitum** [comment]
>  checkout `:h coc#float#has_scroll()`

When trying to scroll as suggested I'm getting lots of `<Ignore><Ignore><Ignore><Ignore><Ignore>` in the buffer, on neovim 0.4.4. My current mapping is:

```vim
inoremap <silent><expr>      <c-j> coc#float#has_scroll() ? coc#float#scroll(1) : "\<c-j>"
inoremap <silent><expr>      <c-k> coc#float#has_scroll() ? coc#float#scroll(0) : "\<c-k>"
vnoremap <silent><expr>      <c-j> coc#float#has_scroll() ? coc#float#scroll(1) : "\<c-j>"
vnoremap <silent><expr>      <c-k> coc#float#has_scroll() ? coc#float#scroll(0) : "\<c-k>"
```

`coc#util#float_scroll_i` has a smoother scroll, so I've changed back to it. I'd suggest to use the argument for scroll to give an amount, instead of direction solely, like done for `coc#util#float_scroll_i`, so users can control how much it scrolls on each step. Maybe check arg is float or number to scroll by window percentage or line count.

**Comment by chemzqm** [comment]
>  I'm getting lots of <Ignore><Ignore><Ignore><Ignore><Ignore> in the buffer, on neovim 0.4.4.

`<Ignore>` is required when we need timer, it works on neovim >= 0.5.0

**Comment by chemzqm** [comment]
I've made some fixes for `coc#float#has_scroll()`, now it works on neovim >= 0.4.3.
The bad news is key-mappings need to be changed.

```
" Note coc#float#scroll works on neovim >= 0.4.3 or vim >= 8.2.0750
nnoremap <nowait><expr> <C-f> coc#float#has_scroll() ? coc#float#scroll(1) : "\<C-f>"
nnoremap <nowait><expr> <C-b> coc#float#has_scroll() ? coc#float#scroll(0) : "\<C-b>"
inoremap <nowait><expr> <C-f> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(1)\<cr>" : "\<Right>"
inoremap <nowait><expr> <C-b> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(0)\<cr>" : "\<Left>"
```

**Comment by oblitum** [comment]
@chemzqm 

I tried it, but it's not behaving well:

- Bad scrollbar for small windows:
  ![imgur-12:07:42_20-10-2020](https://user-images.githubusercontent.com/1269815/96606988-44b1d480-12ce-11eb-9f42-484293a01f51.png)
- Error when leaving signatureHelp, makes coc exit (I call `"quickfixes"` action on `InsertLeave`, if diagnostic changes, I can share the code if needed ~~, this error doesn't happen on release branch~~):
  ```
  Error on request (CocAutocmd): Vim(return):Error invoking 'CocAutocmd' on channel 3 (coc):
  ch 3 returned a response with an unknown request id. Ensure the client is properly synchronized
  Error on request (quickfixes): Vim(return):Error invoking 'quickfixes' on channel 3 (coc):
  [coc] server connection lost
  ```
- Doesn't scroll doc floating window (tried in insert mode):
  ![imgur-12:14:18_20-10-2020](https://user-images.githubusercontent.com/1269815/96607046-54311d80-12ce-11eb-94d0-8ae17d092941.png)

**Comment by oblitum** [comment]
Sorry, closed it by mistake.

---

>  this error doesn't happen on release branch

I want to reword that, I see that release branch was updated, so now all these issues are happening on release branch too. The last working release commit is 9ba22e4aa390bc0b7955352c00f0b76db2a9a016.

**Comment by Frederick888** [comment]
May I know how I can have consistent Ctrl-j/k or S-/Tab mappings for both floating windows (e.g. actions) and pop-ups?

**Comment by oblitum** [comment]
@Frederick888 please ask on gitter as it may derail the topic.

**Comment by chemzqm** [comment]
> May I know how I can have consistent Ctrl-j/k or S-/Tab mappings for both floating windows (e.g. actions) and pop-ups?

Configuration for action menu not available yet, it's recommended to just type number.

**Comment by chemzqm** [comment]
@oblitum I can't reproduce any of your issue on neovim 0.4.4 and neovim 0.4.3, please open new issue with minimal vimrc.

**Comment by Frederick888** [comment]
Not sure if this is related but the completions now don't clean up themselves properly from time to time when using split windows.

![Screenshot_20201021_175517](https://user-images.githubusercontent.com/4507647/96685266-976fb700-13c8-11eb-9c31-d57c14fc6257.png)
![Screenshot_20201021_180638](https://user-images.githubusercontent.com/4507647/96685270-98a0e400-13c8-11eb-83f5-e7b58a70be28.png)

I can open a new issue if needed.

**Comment by chemzqm** [comment]
@Frederick888 it could happen when you're not using neovim 0.5.0 and the float window is not closed by coc.nvim, please provide vim version with minimal vimrc.

**Comment by Frederick888** [comment]
@chemzqm Thanks for the quick response and yup I'm using neovim v0.4.4 so this is probably the reason.

Currently I'm able to reproduce this issue using the minimal vimrc in bug report template.
![image](https://user-images.githubusercontent.com/4507647/96687872-32b65b80-13cc-11eb-802c-a4a3b34359d8.png)
```vim
set nocompatible
set runtimepath^=/home/frederick/.vim/plugged/coc.nvim
filetype plugin indent on
syntax on
set hidden
```

**Comment by Frederick888** [comment]
@chemzqm Actually I just installed neovim v0.5.0 and I still have this problem using the minimal vimrc from bug report template:

![image](https://user-images.githubusercontent.com/4507647/96689030-97be8100-13cd-11eb-8488-58bbf9ff1dec.png)

```
NVIM v0.5.0-765-g75996a2cd
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/gcc-5 -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=auto -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/travis/build/neovim/bot-ci/build/neovim/build/config -I/home/travis/build/neovim/bot-ci/build/neovim/src -I/home/travis/build/neovim/bot-ci/build/neovim/.deps/usr/include -I/usr/include -I/home/travis/build/neovim/bot-ci/build/neovim/build/src/nvim/auto -I/home/travis/build/neovim/bot-ci/build/neovim/build/include
Compiled by travis@travis-job-3016a55c-515c-44fe-9085-50cddd40396e

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/share/nvim"
```

**Comment by chemzqm** [comment]
@Frederick888 I can't reproduce.

**Comment by Frederick888** [comment]
@chemzqm It's been 15 minutes and I've been repetitively doing:

1. Start neovim, split window
2. Trigger a completion
3. Navigate through all options
4. Close neovim

And it happened... once lol. I guess I'll need to get back to you later when I have a stable way to reproduce this.

**Comment by chemzqm** [comment]
@Frederick888 try latest release branch.

**Comment by fannheyward** [comment]
@chemzqm 

<img width="842" alt="WeChat9d79ff21c33254705d1221a6686d69aa" src="https://user-images.githubusercontent.com/345274/96861419-a4f37280-1496-11eb-8cb3-f544d54cb6c9.png">

**Comment by chemzqm** [comment]
@fannheyward shoud be fixed

**Comment by oblitum** [comment]
I'm using this setup to replace the usage of `coc#util#float_scroll_i`:

```vim
let g:coc_snippet_next = '<c-l>'
let g:coc_snippet_prev = '<c-h>'
inoremap <silent><nowait><expr> <c-j> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(1, 1)\<cr>" : "\<c-j>"
inoremap <silent><nowait><expr> <c-k> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(0, 1)\<cr>" : "\<c-k>"
vnoremap <silent><nowait><expr> <c-j> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(1, 1)\<cr>" : "\<c-j>"
vnoremap <silent><nowait><expr> <c-k> coc#float#has_scroll() ? "\<c-r>=coc#float#scroll(0, 1)\<cr>" : "\<c-k>"
```
And it's working great now that the issues I was having are now gone. It seems there's one thing left though, visual mode mappings are not working (either `"\<c-r>=coc#float#scroll(1, 1)\<cr>"` or `coc#float#scroll(1, 1)`). Is there a way to support that, as it works fine on `coc#util#float_scroll_i`? Otherwise it'll miss a killer feature which is to scroll signatureHelp on snippet expansion and jump placeholder.

**Comment by oblitum** [comment]
If there's some way for that to work then I think `coc#util#float_scroll_i` could be simply deprecated or removed, as `coc#float#scroll` can do everything the other can do.

**Comment by oblitum** [comment]
> New input prompt on neovim >= 0.4.3 avoid the need to move your eyes.

I get `:=coc#float#close_i(1004)` in command line after exiting the prompt without changing variable name.

**Comment by chemzqm** [comment]
You can use `<silent>`

**Comment by oblitum** [comment]
@chemzqm it's `<silent>` already, it shows up solely on leaving prompt without changing variable name, which happens much after the mapping.

**Comment by Frederick888** [comment]
@chemzqm With latest release branch and neovim 0.5.0, I can still see those completion window leftovers using my vimrc but not the minimal one. But to properly test this I should probably check out AutoKey etc (it's stopped working under Arch atm...).

But now I see something similar to #2491 almost every time (not using split windows this time):
![Screenshot_20201023_195152](https://user-images.githubusercontent.com/4507647/96982471-9dea6400-156b-11eb-87e9-1ad0981af012.png)

**Comment by oblitum** [comment]
After last commits it seems `"signature.maxWindowHeight"` stopped being respected and signature help windows can become very big.

**Comment by oblitum** [comment]
Must comment that names are not consistent. While you have `signature.maxWindowHeight` for height, there's `signature.floatMaxWidth` for width (this last one is still being respected).

**Comment by Frederick888** [comment]
I've been using it for the last few days and although sometimes it needs a bit patience, the artifacts can be eventually cleaned up. My last comment mentioned that 'can still see those completion window leftovers using my vimrc' but I guess it's simply because I didn't give it enough time.

Thanks for the good work.

**Comment by obxhdx** [comment]
This is a great improvement but the new input prompt is not working for me. The text input is cleared and ignored whenever I press Enter (I'm constantly pressing `<CR>` in the cast below). I'm on Neovim 0.4.4 btw.
[![asciicast](https://asciinema.org/a/2t6BDNSDkI8dbyZYymub2GYQ6.svg)](https://asciinema.org/a/2t6BDNSDkI8dbyZYymub2GYQ6)

Another (minor) thing I noticed is that the prompt title gets truncated if the (eg) the variable you're renaming is at the first line of a buffer:
[![asciicast](https://asciinema.org/a/TTh7RqAk56MMRLy8DqIlxP6v3.svg)](https://asciinema.org/a/TTh7RqAk56MMRLy8DqIlxP6v3)

**Comment by chemzqm** [comment]
>  The text input is cleared and ignored whenever I press Enter 

I can't reproduce, try minimal vimRc.

**Comment by obxhdx** [comment]
Oh, it turns out this only happens when I have [tmsvg/pear-tree](https://github.com/tmsvg/pear-tree) enabled. 🤦‍♂️

**Comment by chemzqm** [comment]
Update: input prompt work on vim as well.

**Comment by VincentCordobes** [comment]
I used to `ctrl-f` to open the command-line window when _renaming_. Is it possible to configure off the new _rename_ floating window to get back the previous behaviour ?

**Comment by chemzqm** [comment]
> I used to `ctrl-f` to open the command-line window when _renaming_. Is it possible to configure off the new _rename_ floating window to get back the previous behaviour ?

coc.preferences.promptInput

**Comment by chemzqm** [comment]
Update: `coc#float#scroll` now support all modes, including visual mode, `coc#float#nvim_scroll` is deprecated.

**Comment by chemzqm** [comment]
Update: add notifications support, not on release branch yet.

**Comment by chemzqm** [comment]
Add `window.withProgress` support, ex:

``` js
  window.withProgress({title: 'Downloading', cancellable: true}, (progress, token) => {
    let n = 0
    return new Promise(resolve => {
      let interval = setInterval(() => {
        progress.report({message: 'progress', increment: 1})
        n = n + 1
        if (n == 100) {
          clearInterval(interval)
          resolve('done')
        }
      }, 100)
      token.onCancellationRequested(() => {
        clearInterval(interval)
        resolve()
      })
    })
  }).then(res => {
    console.log(res)
  })

```

**Comment by ckipp01** [comment]
I feel pretty strongly that the changes made for handling the `window/showMessageRequest` are actually a regression, not an enhancement. In [coc-metals](https://github.com/scalameta/coc-metals) and I'm sure other lsp extensions that make heavy use of the message request, it's very unclear how this actually works. Here are two different popups that I know see:

<img width="1309" alt="Screenshot 2020-12-10 at 11 03 43" src="https://user-images.githubusercontent.com/13974112/101757784-d9b79980-3ad7-11eb-9373-73139077d070.png">

<img width="1309" alt="Screenshot 2020-12-10 at 11 04 11" src="https://user-images.githubusercontent.com/13974112/101757802-dcb28a00-3ad7-11eb-9ddb-862739ec6255.png">

Firstly from a UX perspective, it's not clear at all how you use this. After digging around and asking a bit it also seems that the recommended way for nvim to use this is to use your mouse to select an option. I purposefully don't have `set mouse=a` set in because I don't want to use my mouse. The previous behavior to handle a `window/showMessageRequest` required the user to simply select the option and hit return with no additional settings needed. From my understanding [here on gitter](https://gitter.im/neoclide/coc.nvim?at=5fd1e1ab92aa1c4ef5d002a3) a user would essentially have to create their own function to get a way to bind this to a shortcut. I've already been getting messages today from users that show confusion and frustration about how to use this or how to turn it off.

I think that without having an easy built-in way to select an option here without your mouse is actually a large regression for a Vim user, not an enhancement. Please consider adding an option to either turn this off or provide a simple way for users to select an option that is as easy as it was before.
